### PR TITLE
feat: optimism staking rewards

### PR DESCRIPTION
**Pull request by xgambitox**
## Description

- Add staking rewards use cases on Optimism

## Motivation and Context

- Users should be able to participate in incentivized vaults on Optimism network, by staking their vault tokens to receive OP tokens as reward.

## How Has This Been Tested?

- Tested all use cases on Optimism chain

### CODE REVIEW

File: apps/common/utils/index.tsx [hunk 1]
```diff
@@ -83,3 +83,10 @@ export function	formatDateShort(value: number): string {
 
 	return (new Intl.DateTimeFormat([locale, 'en-US'], {year: 'numeric', month: 'short', day: '2-digit'}).format(value));
 }
+
+/* 🔵 - Yearn Finance **************************************************************************
+**	Returns an object composed of each element of an array, using one of the elements 
+**  properties as its key
+**********************************************************************************************/
+export const keyBy = <T1, T2 extends keyof T1 & string>(array: T1[], key: T2): TDict<T1 | undefined> => 
+	(array || []).reduce((r, x): TDict<T1> => ({...r, [x[key] as string]: x}), {});
```
Comment (karelianpie) [inline apps/common/utils/index.tsx:92]: Looks a bit cryptic, how about having somes tests, GPT suggested the following. Probably missing some edge cases though:

```typescript
import { keyBy } from './keyBy';

type TestObject = {
  id: string;
  name: string;
};

describe('keyBy', () => {
  it('should return an empty object when the input array is empty', () => {
    const input: TestObject[] = [];
    const result = keyBy(input, 'id');
    expect(result).toEqual({});
  });

  it('should create a dictionary keyed by the specified key', () => {
    const input: TestObject[] = [
      { id: '1', name: 'Alice' },
      { id: '2', name: 'Bob' },
      { id: '3', name: 'Charlie' },
    ];
    const expected = {
      '1': { id: '1', name: 'Alice' },
      '2': { id: '2', name: 'Bob' },
      '3': { id: '3', name: 'Charlie' },
    };
    const result = keyBy(input, 'id');
    expect(result).toEqual(expected);
  });

  it('should overwrite duplicate keys with the last occurrence in the array', () => {
    const input: TestObject[] = [
      { id: '1', name: 'Alice' },
      { id: '2', name: 'Bob' },
      { id: '2', name: 'Eve' },
    ];
    const expected = {
      '1': { id: '1', name: 'Alice' },
      '2': { id: '2', name: 'Eve' },
    };
    const result = keyBy(input, 'id');
    expect(result).toEqual(expected);
  });

  it('should return an empty object when the input array is null', () => {
    const input: TestObject[] | null = null;
    const result = keyBy(input as TestObject[], 'id');
    expect(result).toEqual({});
  });

  it('should return an empty object when the input array is undefined', () => {
    const input: TestObject[] | undefined = undefined;
    const result = keyBy(input as TestObject[], 'id');
    expect(result).toEqual({});
  });
});

```

File: apps/vaults/components/RewardsTab.tsx [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import {useAllowances} from '@common/hooks/useAllowances';
+import {useBalance} from '@common/hooks/useBalance';
+import {approveERC20} from '@common/utils/actions/approveToken';
+
+import type {ReactElement} from 'react';
+import type {TAddress} from '@yearn-finance/web-lib/types';
+import type {TYearnVault} from '@common/types/yearn';
+
+const DISPLAY_DECIMALS = 10;
+const trimAmount = (amount: string | number): string => Number(Number(amount).toFixed(DISPLAY_DECIMALS)).toString();
```
Comment (karelianpie) [inline apps/vaults/components/RewardsTab.tsx:22]: Move to utils? Add tests? :)

File: apps/vaults/components/RewardsTab.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+					<div className={'mt-2 text-neutral-600'} >
+						<p>{'Unstake your yVault tokens and your remaining $OP rewards will be claimed automatically. Boom.'}</p>
+					</div>
+				</div>
+				<div className={'flex flex-col gap-4 md:flex-row'}>
+					<Input
+						className={'w-full md:w-[216px]'}
+						label={'You have staked'}
+						legend={formatCounterValue(stakeBalance.normalized, vaultBalance.normalizedPrice)}
+						value={`${trimAmount(stakeBalance.normalized)} ${currentVault.symbol}`}
```
Comment (karelianpie) [inline apps/vaults/components/RewardsTab.tsx:125]: We should probably have a util function for this, `displayAmount`?

File: apps/vaults/components/RewardsTab.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+						isDisabled
+					/>
+					<Button
+						className={'w-full md:mt-7 md:w-[168px]'}
+						onClick={(): unknown => unstake(provider, userAddress, toAddress(stakingRewards?.address))}
```
Comment (karelianpie) [inline apps/vaults/components/RewardsTab.tsx:130]: If we don't have a `stakingRewards?.address`, should we even reach here?

File: apps/vaults/utils/actions/stakingRewards.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+export async function stake(
+	provider: ethers.providers.JsonRpcProvider,
+	accountAddress: TAddress,
+	stakingAddress: TAddress,
+	amount: BigNumber
+): Promise<TTxResponse> {
+	const signer = provider.getSigner(accountAddress);
+	const stakingRewardsContract = new ethers.Contract(stakingAddress, STAKING_REWARDS_ABI, signer);
+	return await handleTx(stakingRewardsContract.stake(amount));
```
Comment (karelianpie) [inline apps/vaults/utils/actions/stakingRewards.tsx:20]: Follow up on https://github.com/yearn/yearn.fi/pull/65/files#r1167806331, as we're not using the returned value there, we could do the following:

```suggestion
	await handleTx(stakingRewardsContract.stake(amount));
	return;
```

File: apps/vaults/utils/actions/stakingRewards.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export async function unstake(
+	provider: ethers.providers.JsonRpcProvider,
+	accountAddress: TAddress,
+	stakingAddress: TAddress
+): Promise<TTxResponse> {
+	const signer = provider.getSigner(accountAddress);
+	const stakingRewardsContract = new ethers.Contract(stakingAddress, STAKING_REWARDS_ABI, signer);
+	return await handleTx(stakingRewardsContract.exit());
```
Comment (karelianpie) [inline apps/vaults/utils/actions/stakingRewards.tsx:30]: Same for this and the ones below

File: apps/vaults/components/RewardsTab.tsx (hunk 2 above)
Comment (karelianpie) [reply apps/vaults/components/RewardsTab.tsx:23]: All good, one last thing, as the function is taking a string, should we add a check for when `isNaN`?

File: apps/vaults/components/RewardsTab.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (karelianpie) [reply apps/vaults/components/RewardsTab.tsx:134]: Let's go with `unknown` for now then :)

**Review by karelianpie — APPROVED** [review]
